application/src/test/java/org/thingsboard/server/service/state/DefaultDeviceStateServiceTest.java

```diff
@@ -341,4 +459,37 @@ private void activityVerify(boolean isActive) {
         Mockito.verify(telemetrySubscriptionService, Mockito.times(1)).saveAttrAndNotify(Mockito.any(), Mockito.eq(deviceId), Mockito.any(), Mockito.eq("active"), Mockito.eq(isActive), Mockito.any());
     }
 
-}
\ No newline at end of file
+    @Test
+    public void givenConcurrentAccess_whenGetOrFetchDeviceStateData_thenFetchDeviceStateDataInvokedOnce() {
+        var deviceStateData = DeviceStateData.builder().build();
+        var getOrFetchInvocationCounter = new AtomicInteger();
+
+        doAnswer(invocation -> {
+            getOrFetchInvocationCounter.incrementAndGet();
+            Thread.sleep(100);
+            return deviceStateData;
+        }).when(service).fetchDeviceStateDataUsingSeparateRequests(deviceId);
+
+        int numberOfThreads = 10;
+        var allThreadsReadyLatch = new CountDownLatch(numberOfThreads);
+        var executor = Executors.newFixedThreadPool(numberOfThreads);
+
+        for (int i = 0; i < numberOfThreads; i++) {
+            executor.submit(() -> {
+                allThreadsReadyLatch.countDown();
+                try {
+                    allThreadsReadyLatch.await();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+                service.getOrFetchDeviceStateData(deviceId);
+            });
+        }
+
+        executor.shutdown();
```

Comment:
In this test, executor is local to this method and because of that I have used try-finally block to ensure that executor will be closed in case something unexpected happens